src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.cs

```diff
@@ -0,0 +1,211 @@
+        // reconnecting.  The connections also cache the prepared statements used
+        // to get/set data from the db.  A connection is safe to use by one thread
+        // at a time, but is not safe for simultaneous use by multiple threads.
+        private readonly object _connectionGate = new object();
+        private readonly Stack<SqlConnection> _connectionsPool = new Stack<SqlConnection>();
+
+        public SQLitePersistentStorage(
+            IOptionService optionService,
+            string workingFolderPath,
+            string solutionFilePath,
+            string databaseFile,
+            Action<AbstractPersistentStorage> disposer)
+            : base(optionService, workingFolderPath, solutionFilePath, databaseFile, disposer)
+        {
+        }
+
+        private SqlConnection GetConnection()
+        {
+            lock (_connectionGate)
+            {
+                // If we have an available connection, just return that.
+                if (_connectionsPool.Count > 0)
+                {
+                    return _connectionsPool.Pop();
+                }
+            }
+
+            // Otherwise create a new connection.
+            return new SqlConnection(this.DatabaseFile);
+        }
+
+        private void ReleaseConnection(SqlConnection connection)
+        {
+            lock (_connectionGate)
+            {
+                // If we've been asked to shutdown, then don't actually add the connection back to 
+                // the pool.  Instead, just close it as we no longer need it.
+                if (_shutdownTokenSource.IsCancellationRequested)
+                {
+                    connection.Close_OnlyForUseBySqlPersistentStorage();
+                    return;
+                }
+
+                _connectionsPool.Push(connection);
+            }
+        }
+
+        public override void Close()
+        {
+            lock (_connectionGate)
+            {
+                // Notify any outstanding async work that it should stop.
+                _shutdownTokenSource.Cancel();
+
+                // Go through all our pooled connections and close them.
+                while (_connectionsPool.Count > 0)
+                {
+                    var connection = _connectionsPool.Pop();
+                    connection.Close_OnlyForUseBySqlPersistentStorage();
+                }
+            }
+        }
+
+        private PooledConnection GetPooledConnection()
+            => new PooledConnection(this, GetConnection());
+
+        public override void Initialize(Solution solution)
+        {
+            // Create a connection to the DB and ensure it has tables for the types we care about. 
+            using (var pooledConnection = GetPooledConnection())
+            {
+                var connection = pooledConnection.Connection;
+
+                // First, create all our tables
+                connection.ExecuteCommand(
```

Comment:
> If the database on disk exists but doesn't have the correct form, it would be good to bail out, delete it, and try again. We should probably also give it a version number internally where a version mismatch triggers a reset (delete and retry).

That's specifically an anti-goal.  We want the DB to be usable across multiple versions of VS as well.

The intent on how you get this is to use different tables in the DB.  i.e. the version info can be encoded into StringTableName/SolutionDataTableName/ProjectDataTableName/DocumentDataTable name.

I can change the current PR to be explicit about what the current version is.